### PR TITLE
Constant Stochastic Gradient Descent

### PR DESCRIPTION
**Pull request by shkr**
Hey,

I recently came across this publication Stochastic Gradient Descent as Approximate Bayesian Inference https://arxiv.org/pdf/1704.04289v1.pdf which I found interesting. 

In comparison to Stochastic Gradient Fisher Scoring which uses a preconditioning matrix to sample
from the posterior even with decreasing learning rates, this work uses optimal constant learning rates such that Kullback-Leibler divergence between the stationary distribution of SGD and the posterior is minimized.

It approximates stochastic variational inference while SGFS and many other MCMC techniques
converge towards the exact posterior. In comparison to SGFS the paper 
discusses a proof for the optimal preconditioning matrix
based on variational inference, so the preconditioning matrix is not an input.

I have implemented it by extending the [BaseStochasticGradient class](https://github.com/pymc-devs/pymc3/blob/master/pymc3/step_methods/sgmcmc.py#L57) introduced in the SGFS PR.

I am submitting this PR before it is complete, to get feedback on this algorithm.

### CODE REVIEW

**Comment by shkr** [comment]
I am unable to understand what is meant by the statement ```We show projections on the smallest and largest principal component of the posterior``` in Figure 1. Any help on how to calculate these projections of a posterior ? 

I want to replicate the results from Fig 1, 2 and 3.

**Comment by twiecki** [comment]
This is a great start @shkr.

To do the PCA projection you simply run SVD on the covariance matrix of the posterior. That gives you M, V, and M.T. You can just take the first and last columns of the V matrix and make that your projection matrix. Let me know if that's not clear.

**Comment by shkr** [comment]
Okay. 
I calculated the covariance of the posterior [ P ] of shape (Q x S) 
Q : number of parameters
S : number of samples in trace

Sigma = { P - mean(P) } * { P - mean(P) }.T of shape (Q x Q)

Then I selected the first and last row from V.h from the SVD decomposition { Sigma = U S V.H }
Afterwards I projected S samples of size (Q x 1) from the trace to the first and last component.

This gives me a S number of 2 dimensional vectors.

Is that what is being done ? I am trying to interpret these projections, but I think I am doing something incorrect. Can you confirm the ^above steps ?

**Comment by shkr** [comment]
^ @twiecki any comments ?

**Comment by twiecki** [comment]
@shkr Sorry, I've been on vacation, will try to take a look soon.

**Comment by twiecki** [comment]
@shkr I think that's close. However, I think they just compute the principle components of the posterior once (e.g. on the NUTS samples) and then project the individual traces  into that.

**Comment by twiecki** [comment]
It does seem to work much better than SGFS, what are your conclusions?

**Comment by fonnesbeck** [comment]
I actually used this as an example of extending PyMC3 in a presentation last week. It worked really well!

**Comment by shkr** [comment]
@twiecki I have updated the notebook. The posterior from the CSG does generate a good approximation for the posterior, figures are falling in line with the paper.  I expected independent projections on the largest and smallest eigenvectors of the sample covariance matrix, however it is not true for SGFS and it is possibly because of two untuned  hyper parameters in SGFS. No such tuning is required for CSG since there are theoretical optimal values for all the hyper parameters. I want to try using CSG for hyper parameter optimization of the lasso model before making final conclusions and request for merge.

**Comment by twiecki** [comment]
@shkr I somewhat forgot about this but I'm fairly excited about the work you've put in here. Do you think we should include the sampler in the code base? Seems like it's preferable over SGFS.

**Comment by shkr** [comment]
@twiecki Yes. I was busy with some other work, so I was unable to push the update here. I will be pushing a commit, this weekend at the latest. It will be ready for review/merge.

**Comment by shkr** [comment]
And yes, I agree the CSG is preferable over SGFS

**Comment by shkr** [comment]
@twiecki @fonnesbeck some debugging help required.

screen shot 2017-11-09 at 10 20 22 pm

screen shot 2017-11-09 at 10 20 30 pm

I am unable to understand, why theano is throwing a disconnected input error here.

As per the model`mu` is the laplacian, and `s` is the regularizer parameter, for the distribution.
I would expect the gradient of the ohs_var which is dependent on mu to have a gradient wrt to `s`.

**Comment by shkr** [comment]
That question is non-blocking to this PR. I just ran into that error while trying to implement the hyper parameter section of the paper. But, having thought about few use cases, it does not make sense, since hyper parameters such as the the number of nodes in a neural network are non-differentiable. So I am a bit unclear, how the EM routine is helpful for general problems.

**Comment by shkr** [comment]
I have updated sgmcmc.py and I have created a new notebook showing usage of ConstantStochasticGradient. This PR is ready for merge.

File: pymc3/step_methods/sgmcmc.py [hunk 1]
```diff
@@ -298,3 +311,98 @@ def competence(var, has_grad):
...
+
+    def __init__(self, vars=None, **kwargs):
+        """
+        Parameters
+        ----------
+        vars : list 
+            Theano variables, default continuous vars
+        kwargs: passed to BaseHMC
+        """
+        super(ConstantStochasticGradient, self).__init__(vars, **kwargs)
```
Comment (twiecki) [inline pymc3/step_methods/sgmcmc.py]: Should add an experimental warning.

**Comment by twiecki** [comment]
Can you also add a note to the `RELEASE-NOTES`?

File: pymc3/step_methods/sgmcmc.py [hunk 2]
```diff
@@ -298,3 +311,98 @@ def competence(var, has_grad):
         if var.dtype in continuous_types and has_grad:
             return Competence.COMPATIBLE
         return Competence.INCOMPATIBLE
+
+
+class ConstantStochasticGradient(BaseStochasticGradient):
```
Comment (junpenglao) [inline pymc3/step_methods/sgmcmc.py]: maybe a shorter name?

Comment (shkr) [reply pymc3/step_methods/sgmcmc.py]: I can change it to CSG, just like what I did with SGFS. Is that okay ?

Comment (twiecki) [reply pymc3/step_methods/sgmcmc.py]: Yeah, it's not great but for consistency probably the best option.

File: pymc3/step_methods/sgmcmc.py (hunk 1 above)
Comment (shkr) [reply pymc3/step_methods/sgmcmc.py]: The experimental warning is present in this line https://github.com/shkr/pymc3/blob/45a45ab0f78b480b8accb27f168164bb213cd280/pymc3/step_methods/sgmcmc.py#L112

File: pymc3/step_methods/sgmcmc.py (hunk 2 above)
Comment (shkr) [reply pymc3/step_methods/sgmcmc.py]: I have renamed it

**Comment by shkr** [comment]
@twiecki I have inserted a line in the RELEASE-NOTES and added my name as a community member. Let me know if thats what you wanted.

File: RELEASE-NOTES.md [hunk 3]
```diff
@@ -249,6 +250,7 @@ Patricio Benavente <patbenavente@gmail.com>
 Raymond Roberts
 Rodrigo Benenson <rodrigo.benenson@gmail.com>
 Sergei Lebedev <superbobry@gmail.com>
+Shashank Shekhar <shashank.f1@gmail.com>
```
Comment (twiecki) [inline RELEASE-NOTES.md]: Can you add yourself instead to a new section `Contributors` for the upcoming 3.3 release?

File: pymc3/step_methods/sgmcmc.py [hunk 4]
```diff
@@ -298,3 +311,99 @@ def competence(var, has_grad):
...
+class CSG(BaseStochasticGradient):
+    R"""
+    CSG: ConstantStochasticGradient
+    
+    It is an approximate stochastic variational inference algorithm
+    while SGFS and many other MCMC techniques provably
+    converge towards the exact posterior. The referenced paper 
+    discusses a proof for the optimal preconditioning matrix
+    based on variational inference, so there is no parameter tuning required
+    like in the case of 'B' matrix used for preconditioning in SGFS.
```
Comment (twiecki) [inline pymc3/step_methods/sgmcmc.py]: Maybe link people to your Notebook.

**Comment by twiecki** [comment]
Can you also add the NB to the docs? And make sure you only have one top-level heading `#` in the NB.

**Comment by shkr** [comment]
@twiecki Done ! I added the 3 notebooks I have created for stochastic algorithms to the examples doc.

File: RELEASE-NOTES.md (hunk 3 above)
Comment (shkr) [reply RELEASE-NOTES.md]: Done

**Comment by junpenglao** [comment]
Great job @shkr! Just a nitpick: in constant_stochastic_gradient.ipynb you still have

> top-level heading `#`

at the end. You should do `# Result` --> `## Result`

**Comment by shkr** [comment]
@junpenglao done !

**Comment by twiecki** [comment]
```
************* Module pymc3.sampling

pymc3/sampling.py:13: [W0611(unused-import), ] Unused CSG imported from step_methods

pymc3/sampling.py:13: [W0611(unused-import), ] Unused SGFS imported from step_methods
```

Would also be curious how CSG does on the neural network, but this doesn't have to be part of this PR.

**Comment by shkr** [comment]
@twiecki Yes. I agree. I will put up a follow up PR, with CSG on the neural net and other notebook updates to the stochastic gradient docs.

File: RELEASE-NOTES.md [hunk 5]
```diff
@@ -7,7 +7,8 @@
  
 - Improve NUTS initialization `advi+adapt_diag_grad` and add `jitter+adapt_diag_grad` (#2643) 
 - Update loo, new improved algorithm (#2730)
-
+- New CSG (Constant Stochastic Gradient) approximate posterior sampling
+  algorithm added
```
Comment (twiecki) [inline RELEASE-NOTES.md]: Link to PR like above.

File: docs/source/examples.rst [hunk 6]
```diff
@@ -78,3 +78,12 @@ Variational Inference
    notebooks/convolutional_vae_keras_advi.ipynb
    notebooks/empirical-approx-overview.ipynb
    notebooks/normalizing_flows_overview.ipynb
+   
+
+Stochastic Gradient 
+=====================
```
Comment (twiecki) [inline docs/source/examples.rst]: extra `==`

File: pymc3/step_methods/sgmcmc.py [hunk 7]
```diff
@@ -298,3 +312,102 @@ def competence(var, has_grad):
...
+    Take a look at this example notebook 
+    https://github.com/pymc-devs/pymc3/tree/master/docs/source/notebooks/constant_stochastic_gradient.ipynb
+     
+    Parameters
+    -----
```
Comment (twiecki) [inline pymc3/step_methods/sgmcmc.py]: Make line as long as text above it

**Comment by twiecki** [comment]
Thanks @shkr, this is a significant contribution!

**Comment by twiecki** [comment]
Just tried running this on a larger NN but `output, _ =  theano.scan(lambda i, logX=logL, v=var: theano.grad(logX[i], v).flatten(),\
---> 56                            sequences=[tt.arange(logL.shape[0])])` seems to take forever. Is there no way to vectorize this?
